### PR TITLE
wlink: init at 0.0.9

### LINKED ISSUE #346154: Package request: wlink
**Project description**

An open source WCH-Link library/command line tool written in Rust. It's for connecting/flashing ch32v RISC-V microcontrollers.

This is similar to the wch-isp package. @LeSuisse, as the maintainer of wch-isp, would you be interested in packaging this? Otherwise, I can take a stab at it.

**Metadata**

* homepage URL: https://github.com/ch32-rs/wlink
* source URL: https://github.com/ch32-rs/wlink
* license: Dual Apache 2.0 and  MIT
* platforms: Unix, Darwin

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by jwillikers**
Add [wlink](https://github.com/ch32-rs/wlink) package. wlink is an open source WCH-Link library/command line tool written in Rust. It's for connecting/flashing ch32v RISC-V microcontrollers.

Fixes #346154.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/wl/wlink/package.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    updateScript = nix-update-script { };
+    tests.version = testers.testVersion {
+      package = wlink;
+      version = "${version}";
+    };
+  };
+
+  meta = with lib; {
+    description = "WCH-Link flash tool for WCH's RISC-V MCUs(CH32V, CH56X, CH57X, CH58X, CH59X, CH32L103, CH32X035, CH641, CH643)";
+    homepage = "https://crates.io/crates/wlink";
```
Comment (h7x4) [inline pkgs/by-name/wl/wlink/package.nix:41]: No big deal about the homepage, the github repo just seems a better landing page to explore the project. The changelog would be nice though

```suggestion
    homepage = "https://github.com/ch32-rs/wlink";
    changelog = "https://github.com/ch32-rs/wlink/releases/tag/v${version}";
```

File: pkgs/by-name/wl/wlink/package.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    license = with licenses; [
+      mit # or
+      asl20
+    ];
+    platforms = platforms.linux;
```
Comment (h7x4) [inline pkgs/by-name/wl/wlink/package.nix:46]: Upstream seems to support more than just linux, we'll just mark the rest as broken for now.

```suggestion
    platforms = with platforms; linux ++ darwin ++ windows;
    broken = !stdenv.hostPlatform.isLinux;
```

File: pkgs/by-name/wl/wlink/package.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  buildInputs = [
+    libusb1
+    udev
+  ];
+
+  passthru = {
+    updateScript = nix-update-script { };
+    tests.version = testers.testVersion {
+      package = wlink;
+      version = "${version}";
```
Comment (h7x4) [inline pkgs/by-name/wl/wlink/package.nix:35]: This is the default AFAIK

```suggestion
```

**Review by h7x4 — COMMENTED** [review]
Verified github id and gpg key :+1:

File: pkgs/by-name/wl/wlink/package.nix (hunk 3 above)
Comment (jwillikers) [reply pkgs/by-name/wl/wlink/package.nix]: Good to know, thanks!

File: pkgs/by-name/wl/wlink/package.nix (hunk 1 above)
Comment (jwillikers) [reply pkgs/by-name/wl/wlink/package.nix]: Yeah, that does seem like a better landing page.

File: pkgs/by-name/wl/wlink/package.nix (hunk 2 above)
Comment (h7x4) [reply pkgs/by-name/wl/wlink/package.nix]: Ah forgot to mention, remember to add `stdenv` to the arguments up top

**Comment by h7x4** [comment]
Could you squash the fixups into the commit that adds the package? Just be careful not to rebase past commits, it tends to cause a mass ping that will require creating a new PR. If you're unsure, it could be helpful to draft the PR first - it will mitigate a mass ping despite an erroneous rebase.

**Comment by jwillikers** [comment]
> Could you squash the fixups into the commit that adds the package? Just be careful not to rebase past commits, it tends to cause a mass ping that will require creating a new PR. If you're unsure, it could be helpful to draft the PR first - it will mitigate a mass ping despite an erroneous rebase.

No problem, I thought that might be preferred.

**Review by h7x4 — APPROVED** [review]

**Pull request merged**